p2p/switch.go

```diff
@@ -681,6 +681,9 @@ func (sw *Switch) addPeer(p Peer) error {
 	}
 	sw.metrics.Peers.Add(float64(1))
 
+	for _, reactor := range sw.reactors {
+		p = reactor.InitPeer(p)
+	}
```

Comment:
![image](https://user-images.githubusercontent.com/7310198/55306833-25531c80-5488-11e9-9da2-dacea5cc81fe.png)

Reply:
keyPoint is :  there is not handshake that tell other peer I am ready to receive message. So do some init work before the mconnection start.

Reply:
If I understand correctly is that the peer's mconnection is started too early, but the introduction of `InitPeer` won't help with that as [`p.Start()`](https://github.com/tendermint/tendermint/blob/e82f69babb977655751fbcac168cebe5daa94bc3/p2p/switch.go#L669) is called befpre that.

The only sensible way forward is to find a way to write a failing test for this scenario and then verify that any proposed solution solves.

Reply:
I tend to agree with @xla here, not clear the fix will work. Curious, have you tried to actually delay the peer reconnecting? (e.g. in switch.go:reconnectToPeer() bring the sw.randomSleep() just at the beginning of the first for loop.)
My understanding is that the root cause of the issue is the peer reconnecting too fast, before RemovePeer() has a chance to cleanup. I think it would be ok to delay the reconnection of a peer that was stopped for error.

Reply:
> My understanding is that the root cause of the issue is the peer reconnecting too fast, before RemovePeer() has a chance to cleanup. 

**the core issue here** is that `reactor.Receive` might be called before `reactor.AddPeer`, which will lead to reactor panicking if it depends on data from`reactor.AddPeer`. **The proposed solution** is to move data initialization to a new `reactor.InitPeer` function, which is guaranteed to be executed before `reactor.Receive` is called.